brainframe_qt/ui/resources/video_items/streams/stream_listener_widget.py

```diff
@@ -88,61 +53,72 @@ def handle_stream_url(stream_url: Optional[str]) -> None:
             if stream_url is None:
                 return
 
+            # User must have already changed stream again by the time this callback is
+            # called. Just forget about this current change request.
+            if stream_conf is not self.stream_conf:
+                return
+
             self._subscribe_to_stream(stream_conf, stream_url)
 
         QTAsyncWorker(self, self._get_stream_url, f_args=(stream_conf,),
                       on_success=handle_stream_url) \
             .start()
 
-    def _clear_current_stream_reader(self):
-        """If we currently have a stream reader, unsubscribe its listener
-        and clear any posted events"""
-
-        # Ensure that we're not storing a stream_conf
-        self.stream_conf = None
-
-        if not self.stream_reader:
-            return
-
-        self.destroyed.disconnect()
-
-        self.stream_reader.remove_listener(listener=self)
-
-        # Make sure no more events are sent to this listener
-        QCoreApplication.removePostedEvents(self)
+    def _disconnect_stream_reader(self) -> None:
+        # Disconnect existing signals
+        # This will be unnecessary when composition is used with StreamListener instead
+        # of inheritance and PyQt manages this during GC.
+        if self.stream_reader is not None:
+            self.stream_reader.frame_received.disconnect(self.on_frame)
+            self.stream_reader.stream_state_changed.disconnect(self._on_state_change)
 
-        self.stream_reader = None
+    def _on_state_change(self, state: StreamStatus) -> None:
+        if state is StreamStatus.INITIALIZING:
+            self.on_stream_init()
+        elif state is StreamStatus.HALTED:
+            self.on_stream_halted()
+        elif state is StreamStatus.CLOSED:
+            self.on_stream_closed()
+        elif state is StreamStatus.STREAMING:
+            # Streaming, but no frame received yet
+            self.on_stream_init()
+        else:
+            self.on_stream_error()
```

Comment:
This can be removed in a future PR by having widgets directly listen to the `stream_state_changed` signal. Right now `StreamListener` is used through inheritance instead of composition, and I would like to change that too. But isn't strictly necessary for this PR.